moveit_ros/moveit_servo/src/servo_node.cpp

```diff
@@ -296,9 +298,12 @@ void ServoNode::servoLoop()
 
   while (rclcpp::ok() && !stop_servo_)
   {
-    // Skip processing if servoing is disabled.
-    if (servo_paused_)
+    // Skip processing commands if servoing is disabled, but allow parameter updates.
```

Comment:
I understand not changing the planning group unless servo is paused... but for all the params? Like, if I want to change a scaling factor or enable/disable collision it should now be paused?

Reply:
Yeah, this makes things more safer (updating parameters while the control is running doesn't sound safe) and also efficient since update checks are skipped when servo is not paused.

@AndyZe  suggested it during the discussions here.

Reply:
Does it count as paused when the robot is publishing nothing with no recent input messages, i.e., smoothHalt() is completed?

That seems like a good option vs. Asking uses to explicitly call 2 ROS services before and after changing any param.

Reply:
Also... what does this mean if you update parameters while servo is unpaused, and then pause/resume? Will you not hit the params out of date flag and have your params be out of sync forever going forward?

I agree it's a good idea for the robot to be stopped before changing values, but we just need to think it through a little to not introduce weird bugs/behaviors.

Reply:
> That seems like a good option vs. Asking users to explicitly call 2 ROS services before and after changing any param.

But there is no way to enforce that updates don't happen during the control cycle, I guess we can just add a note about this.

> Also... what does this mean if you update parameters while servo is unpaused, and then pause/resume? Will you not hit the params out of date flag and have your params be out of sync forever going forward?

Yup, I hit this issue while testing.

Reply:
Yeah, thinking about this, the logic complexity may get a bit complicated.

I would say for the core Servo API, users can do whatever they want and we should add notes about the risks of switching planning groups while moving.

For the servo node, we could still call `updateParameters()` as normal, but keep the "next planning group" stashed in a member variable until Servo has been paused or is otherwise not publishing any commands.

Reply:
I haven't fully thought about this, but I would say the possibility of desyncing the parameters if not paused is a non-starter for me.

Reply:
> For the servo node, we could still call `updateParameters()` as normal, but keep the "next planning group" stashed in a member variable until Servo has been paused or is otherwise not publishing any commands.

Not sure about stashing it, since it would take effect only the next time servo is paused ,which doesn't feel intuitive.

Makes me want to go back to having planning group switch as a service since it would take care pausing , updating and unpausing and the working is very clear.

Reply:
I think it's a good idea to explicitly require the user to pause Servo before changing any param. No risk of desyncing then, right?

Pretty much agree with what Ibrahim said here:

> this makes things more safer (updating parameters while the control is running doesn't sound safe) and also efficient since update checks are skipped when servo is not paused.

---

> For the servo node, we could still call updateParameters() as normal, but keep the "next planning group" stashed in a member variable until Servo has been paused or is otherwise not publishing any commands.

Sounds complicated

Reply:
I guess we could add one exception for the collision checking flag, that it's allowed to be updated while unpaused

Reply:
> I think it's a good idea to explicitly require the user to pause Servo before changing any param. No risk of desyncing then, right?

We will still have a desync issue since the parameters would be updated in the parameter server. We can only enforce `Servo` not updating it's stored parameters.

Reply:
Yeah... The service call option could be promising.

I think something that drives the robot to a halt, pauses, switches planning groups, restarts the smoothing plugin, and resumes would be safe.

---

> I guess we could add one exception for the collision checking flag, that it's allowed to be updated while unpaused

My understanding is that most servo parameters can/should be updated while unpaused, and things like the planning group are the exceptions?

Reply:
But why? Is anybody asking for this? (Other than collision checking toggle)

Definitely anything timing or filtering-related shouldn't be touched while unpaused

Reply:
On MoveIt Studio we currently modify the collision check toggle and the velocity scaling parameters at runtime -- admittedly, not the others.

I guess my point of view here is servo has a ton of gains/thresholds that would make sense to have runtime tunable because why not? The ROS parameter tooling basically supports this for free.

So I wouldn't want to immediately sacrifice that and take a step back to see how we could handle the planning group switching in a way that doesn't revert a lot of the benefits brought on by switching to `generate_parameter_library`.

Reply:
i think we're starting to go in circles. Switching the planning group while unpaused isn't a good idea b/c it will screw up filtering, for one issue

Reply:
100% agree. I just don't want to introduce bugs or kill all our other runtime parameter tunability to singularly satisfy that requirement.

Reply:
How about we allow update of simple parameters like collision toggle etc normally, and check against servo paused status for just planning group change. If not paused we can set status of servo to invalid and print a warning, setting the status to invalid will also skip rest of computations in `getNextJointState` and robot would  be send a command .

Reply:
The above method would also avoid the desync issue since parameters of servo would be updated but since status is invalid, no computations will happen until servo was paused and the update is applied properly. It will keep printing warnings till then.

Reply:
hmm, good point. A warning would go a long way -- then it's user error :)

Reply:
@ibrahiminfinite that sounds OK.

Can we default nearly all of the parameters to static initially, please? We can unlock them one-by-one as people request it and they get carefully considered. For example, here's a list of parameters that should definitely be static. There may be more.

> monitored_planning_scene_topic
> status_topic
> command_out_topic
> command_out_type
> cartesian_command_in_topic
> joint_command_in_topic
> joint_topic
> thread_priority
> incoming_command_timeout
> publish_period
> use_smoothing
> smoothing_filter_plugin_name

check_collisions and scale can be "unlocked" right now for dynamic updates.

Reply:
This seems like a good way forward.

We have a MoveIt Studio configuration with UR arm on a linear rail with multiple planning groups (arm+rail vs. rail only), so I'll be happy to test on that once we're at that point.

Wouldn't be a bad idea to use/introduce a similar configuration with multiple planning groups in open-source MoveIt to actually test/showcase these capabilities.